### ISSUE TITLE
Rule Proposal: nonblock-body-position

### ISSUE THREAD

**Issue by mysticatea** [eslint/eslint#6067]
From [requireNewlineBeforeSingleStatementsInIf](http://jscs.info/rule/requireNewlineBeforeSingleStatementsInIf).

This rule will enforce the position of bare statements at the body of `if`, `else`, `do-while',`while`,`for`,`for-in`, or`for-of`.

``` js
{
    "nonblock-body-position": ["error",  "below" or "beside" or "any"],
    // or
    "nonblock-body-position": ["error", {
        "position": "below" or "beside" or "any",
        "overrides": {
            "if": null,
            "else": null,
            "do-while": null,
            "while": null,
            "for": null,
            "for-in": null,
            "for-of": null
        }
    }]
}
```
- `position`
  - `"below"` (default) - Requires newline before the body's statement.
  - `"beside"` - Disallow newline before the body's statement.
  - `"any"` - Does not enforce the position.
- `overrides` - People can override setting for each kind.

**Valid:**

``` js
/*eslint nonblock-body-position: ["error", "below"]*/

if (a)
    foo();
else
    foo();

while (a)
    foo();

// ignore block statements (those are waned by brace-style)
if (a) { foo(); }
if (a) {
    foo();
}
```

``` js
/*eslint nonblock-body-position: ["error", "beside"]*/

if (a) foo();
else foo();

while (a) foo();

// ignore block statements (those are waned by brace-style)
if (a) { foo(); }
if (a) {
    foo();
}
```

``` js
/*eslint nonblock-body-position: ["error", {"position": "beside", "overrides": {"while": {"position": "below"}}}]*/

if (a) foo();
else foo();

while (a) 
    foo();
```

**Invalid:**

``` js
/*eslint nonblock-body-position: ["error", "below"]*/

if (a) foo();
else foo();

while (a) foo();
```

``` js
/*eslint nonblock-body-position: ["error", "beside"]*/

if (a)
    foo();
else
    foo();

while (a)
    foo();
```

``` js
/*eslint nonblock-body-position: ["error", {"position": "beside", "overrides": {"while": {"position": "below"}}}]*/

if (a)
    foo();
else
    foo();

while (a) foo();
```

**Comment by ilyavolodin** [comment]
I think this rule is going to conflict with `brace-style` `allowSingleLine` option.

**Comment by mysticatea** [comment]
I tried it, but `brace-style` does not check non-block statements.

**Comment by danielcavanagh** [comment]
the other problem with `allowSingleLine` is that it can't **enforce** the use of a single line, so `--fix` does nothing

it seems that positioning of the single line, whether braced or not, should be part of this one rule (perhaps renamed to `single-statement-body-position`?) and the ruling on whether to brace should be left to `curly` (which should probably also be renamed...)

**Comment by mysticatea** [comment]
Thank you @danielcavanagh . 
I had misunderstood it. I have read tests of `requireNewlineBeforeSingleStatementsInIf`, then it's warning statements in a block as well.

Also, original rule is warning the first statement of a block which has multiple statements as well.

**Comment by mysticatea** [comment]
So, the intention of original JSCS rule is "Disallow that any statements locate at beside IfStatements.".

For JSCS compatibility:

- We can use `{allowSingleLine: false}` option of `brace-style` rule if there is a block.
- We can use this rule if there is not a block.

**Comment by kaicataldo** [comment]
How does this relate to https://github.com/eslint/eslint/pull/8099 now that we have that hammered out? Since we're consolidating some other existing rules that enforce padded newlines around nodes into that rule, could this conflict? Should this maybe be covered by that rule as well?

**Comment by not-an-aardvark** [comment]
As far as I can tell, this addresses a separate issue from #8099. #8099 addresses newlines between *consecutive* statements, whereas this addresses newlines between *nested* statement nodes. For example, consider the following `newline-between-statements` configuration:

```js
/* eslint newline-between-statements: ["error", ["blankline", "if", "return"]] */

if (foo)
  return 1; // newline-between-statements *does not* enforce a newline before this statement.

return 2; // newline-between-statements *does* enforce a newline before this statement.
```

**Comment by kaicataldo** [comment]
@not-an-aardvark Thanks, that's a really helpful distinction. Makes sense!

**Comment by kaicataldo** [comment]
Sorry to respond to this so late in the game, but I also wonder if we can decide on a name that more clearly describes what this rule is checking. `nonblock-body-position` wasn't immediately clear to me, so I can imagine it would be confusing to someone who's not a project maintainer. I know it's long, but I think `nonblock-statement-body-position` would be clearer.

**Comment by mysticatea** [comment]
`nonblock-statement-body-position` makes sense to me.

I look at this again, I wonder if `nonblock-statement-body-position` should include the body of arrow expression as well, or not.

```js
() => foo()
// or
() =>
    foo()
```

**Comment by kaicataldo** [comment]
@mysticatea Thinking about this more, I wonder if we shouldn't include ArrowFunctionExpressions. Seems like it's the odd one out, since the rest of the nodes it checks are statements. Could be confusing, particularly if we change the name to include `statements`. I think I'm actually :-1: to including them.

**Comment by platinumazure** [comment]
We can always cover arrow function expressions in a separate rule. I think we should focus on statements in this one.

**Comment by mysticatea** [comment]
About including arrow expressions seems unpopular in team, so let's go as is.
